### PR TITLE
Alignment updates

### PR DESCRIPTION
**Pull request by siilieva**
SciFi and MuFi spatial alignment fit update - switching from Gaussians to Crystal Ball fits
Update the MuFilter spatial alignment code to reflect new detector configuration.
Reduce the number of loops the script that modifies the geofile(modifyGeoFile.py) using dictionaries, add support for 2024 detector setup.

### CODE REVIEW

File: shipLHC/scripts/Survey-MufiScifi.py [hunk 1]
```diff
@@ -3560,31 +3563,41 @@ def Scifi_residuals(Nev=options.nEvents,NbinsRes=100,xmin=-2000.,alignPar=False,
...
                 globalPos[p+projs[o]].SetMarkerStyle(21)
                 globalPos[p+projs[o]].SetMarkerColor(ROOT.kBlue)
             if proj == 'C':
                 for m in range(nMats):
                      h[hname+str(m)] = h[hname].ProjectionX(hname+str(m),m*512,m*512+512)
-                     rc = h[hname+str(m)].Fit('gaus','SQ0')
-                     fitResult = rc.Get()
+                     mu = ROOT.RooRealVar("mu", "residual [cm]", h[hname+str(m)].GetMean(), h[hname+str(m)].GetMean()-200, h[hname+str(m)].GetMean()+200)
+                     width = ROOT.RooRealVar("width", "width", h[hname+str(m)].GetRMS(), 1e-5,  h[hname+str(m)].GetRMS()+1e+3)
+                     dcbPdf = ROOT.RooCrystalBall("dcbPdf", "DoubleSidedCB", x, mu, width, a1, p1, True)
```
Comment (olantwin) [inline shipLHC/scripts/Survey-MufiScifi.py:3645]: Ditto here.

It might be worth comparing with two independent tails.

File: shipLHC/modifyGeoFileDict.py [hunk 2]
```diff
@@ -12,16 +12,20 @@
...
 '''
-supportedGeoFiles = ["geofile_sndlhc_TI18_V7_22November2022.root"]
+supportedGeoFiles = ["geofile_sndlhc_TI18_V7_22November2022.root", "geofile_sndlhc_TI18_V0_2024.root"]
             
-def modifyDicts():
+def modifyDicts(year=2024):
    for g in supportedGeoFiles:
-         os.system('xrdcp -f '+os.environ['EOSSHIP']+'/eos/experiment/sndlhc/convertedData/commissioning/TI18/'+g+' '+g)
+         if year == 2024 and g.find(str(year))>0:
+           os.system('xrdcp -f '+os.environ['EOSSHIP']+'/eos/experiment/sndlhc/convertedData/physics/2024/'+g+' '+g)
```
Comment (olantwin) [inline shipLHC/modifyGeoFileDict.py:20]: Not a fan of this method. It irks me that it depends on:

1. an environment variable (I'd prefer to just give the explicit path, which is known)
2. an external command, called via `os.system`.
3. The `'+g+' '+g` stuff could not be more cryptic if that were the goal.

Using ROOT I/O or xroot directly via its python interface might be better as there are fewer things that could go wrong.

The idea is to cache the geofile locally?`

File: shipLHC/modifyGeoFileDict.py [hunk 3]
```diff
@@ -204,8 +211,29 @@ def modifyDicts():
    0.00*u.mrad, 0.00*u.mrad, 0.00*u.mrad,
    0.00*u.mrad, 0.88*u.mrad, 0.00*u.mrad,
    0.00*u.mrad, 0.07*u.mrad, 0.00*u.mrad]
-   
-         for c in ['t_0', 't_4361','t_4575','t_4855','t_5172','t_5431', 't_6305']:
+# 2024
+         alignment['t_7648']=[       # 2024 emuslsion run 8
+   509.86*u.um, 732.95*u.um, 621.19*u.um,
```
Comment (olantwin) [inline shipLHC/modifyGeoFileDict.py:216]: What's going on with the indentation here?

**Review by olantwin — CHANGES_REQUESTED** [review]
Minor comments, feel free to ignore.

**Comment by olantwin** [comment]
I generally appreciate the slight modernisations and un-hardcoding of magic numbers!

File: shipLHC/scripts/Survey-MufiScifi.py (hunk 1 above)
Comment (siilieva) [reply shipLHC/scripts/Survey-MufiScifi.py:3645]: What would be the (physics/geometry/detector) reason for different tails on the left and right?
The residual is hit - track coordinates at the same Z.

Comment (olantwin) [reply shipLHC/scripts/Survey-MufiScifi.py:3645]: Distortion/non-planarity? Assymetric effects are most likely negligible, but it would be interesting to test whether we are sensitive.

As a baseline for the alignment symmetric tails should be fine, and I don't suggest doing the alignment with two independent tails by default!

File: shipLHC/modifyGeoFileDict.py (hunk 3 above)
Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:216]: for the lines with constants, I am copy-pasting print messages from the alignment script and it is setup to this format. So I go along.
For the comments,  I find it easier to spot where to add the new constants with the comments being more/less indented that the blocks of constants in the lists. I tried using long comment lines (lines of ---- or ####)  to separate the different sets, but it was very distracting since the blocks with constants are rather short(~5 lines each)
In a nutshell, I prefer the keep this file ugly but safe and convenient to my usage.

File: shipLHC/modifyGeoFileDict.py (hunk 2 above)
Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:20]: 1. the idea is that the script is aware of only the baseline geofiles that  have the nominal detector configuration with all alignment constants set to 0. On top of that geometry we add the calibrations. Since we have many more geofiles on eos( with target exchanges we generate updated versions of the geofile), it is good to have path/geo_filename already in the code and not let users change it.
 2. we need the os.system command if the code is run on a local machine. What else can be used to support the remote access?
 3. ok `g` can be changed to `geofile_name`

With all files that get modified, including the geofile, the procedure is to copy it locally, adapt and
either back it up and provide the new version on eos automatically or stop the current code, change the name and copy to EOS manually.
With the geofile we go the second way since we increment the tag (VX) to the geofile name.

File: shipLHC/scripts/Survey-MufiScifi.py (hunk 1 above)
Comment (siilieva) [reply shipLHC/scripts/Survey-MufiScifi.py:3645]: I think the multiple scattering will cover-up these effects by far.  We have testbeam data with no W between SciFi stations and the CB with asymmetric tails can be tested there, out of curiosity even

File: shipLHC/modifyGeoFileDict.py (hunk 2 above)
Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py:20]: 1. The point here is rather, that we just should give the full path including the EOS server. The only effect of our approach is that we might have strange effects if EOSSHIP is ever not set.
2. As I said, ROOT I/O or XRootD via its python bindings.

Thanks for the explanation of the use case.

Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:20]: ok, now that I understand the meaning of point 1, I agree with the suggestion.
On the point 2, do you mean to use like `server_xxx.copy("remote_geofile", "local_copy_geofile")` in place of the` os.system('xrdcp ...')`?
Why is that preferred? Do you have a link where it is explained?

Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py:20]: 2. Yes, basically. With ROOT I/O, the best might be to make a new file and then copy the contents of the old (or only stream the old and immediately write the new). With the xrd client, it's something we'd need to look up.
The reason this would be preferred, would be because we skip calling a separate process via the shell (not sure, whether the shell is also new), which again adds things that can go wrong.

Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py]: You can use `os.getcwd()` or `Path.cwd()`

Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py]: > You can use `os.getcwd()` or `Path.cwd()`

ah, I thought `os.getcwd()` calls the` os.system` in the back, now I checked it is not the case. 
Ok, so I change to the `server.copy(...)` and  use the `os.getcwd()`. Will push the new stuff in a few mins.

File: shipLHC/modifyGeoFileDict.py [hunk 4]
```diff
@@ -4,24 +4,33 @@
...
+
+def modifyDicts(year=2024):
+   for geoFileName in supportedGeoFiles:
+         if (year == 2024 and geoFileName.find(str(year)) > 0) or \
+            (year != 2024 and geoFileName.find('2022') > 0):
+            # override locally existing file with the same name
+            status = theClient.copy(supportedGeoFiles[geoFileName]+geoFileName,
+                                    os.getcwd()+'/'+geoFileName, force=True)
+            if not status[0].ok: 
+               print('Bad status of the local copy process', status)
```
Comment (olantwin) [inline shipLHC/modifyGeoFileDict.py:27]: Should the script abort in this case?

**Review by olantwin — CHANGES_REQUESTED** [review]
Is it worth continuing if the copy fails?

Everything else looks good now!

File: shipLHC/modifyGeoFileDict.py (hunk 4 above)
Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:27]: it crashes if the file is not copied in any case, printing error messages including _file not found_. I added the print of the copy status since it can be useful to identify the copy-related issue

Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py:27]: In line 29?
Might still be worth just exiting here with a failed return code (or raise an exception, OSError/IOError might be appropriate)

Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:27]: yes, line 29, when trying to open the non-existing local copy. 
As long as the code _exists_ and the _useful debug info is there,_ what does it matter precisely how the code handles the error?

Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py:27]: Not really, but imho the cleanest solution would be:

```python
raise RuntimeError('Bad status of the local copy process' + str(status))
```

replacing the print.

Comment (olantwin) [reply shipLHC/modifyGeoFileDict.py:27]: I think if we just let ROOT crash, the problem will be that some people might think it's an issue with a local file not existing, rather than the copy going wrong.

Then again, if we will just get an email in either case, it doesn't really matter. It seems like most of our collaborators don't actually read error messages anyway :D

Comment (siilieva) [reply shipLHC/modifyGeoFileDict.py:27]: this code is used/run by the responsible for the geometry (software coordinator in our case), so they should read the full error log and make sense of it :D
I propose to leave the code as it is.
In general, to make our error logs more 'inviting' and 'pushing' for the users to look at, we can use colors, CAPS etc. I will take a look at that for next code modifications. In case you have something in mind in that direction, do share :)

**Review by olantwin — APPROVED** [review]
